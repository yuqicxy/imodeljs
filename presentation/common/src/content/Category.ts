/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Content
 */

/**
 * A data structure that describes a [[Field]] category.
 * @public
 */
export interface CategoryDescription {
  /** Unique name */
  name: string;
  /** Display label */
  label: string;
  /** Extensive description */
  description: string;
  /** Priority. Categories with higher priority should appear higher in the UI */
  priority: number;
  /** Should this category be auto-expanded when it's displayed in the UI */
  expand: boolean;
}
