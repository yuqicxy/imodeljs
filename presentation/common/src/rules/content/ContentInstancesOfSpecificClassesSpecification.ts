/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */

import { ContentSpecificationBase, ContentSpecificationTypes } from "./ContentSpecification";
import { MultiSchemaClassesSpecification } from "../ClassSpecifications";

/**
 * Creates content for ECInstance(s) of specified classes.
 *
 * **Note**: this specification doesn't rely on selection. It always returns instances no matter
 * what the selection is, so pre-filtering should be done in [[ContentRule]] condition and [[instanceFilter]].
 *
 * @public
 */
export interface ContentInstancesOfSpecificClassesSpecification extends ContentSpecificationBase {
  /** Used for serializing to JSON. */
  specType: ContentSpecificationTypes.ContentInstancesOfSpecificClasses;

  /**
   * List of ECClass specifications whose ECInstances should be selected.
   */
  classes: MultiSchemaClassesSpecification | MultiSchemaClassesSpecification[];

  /**
   * Should all [[classes]] be treated polymorphically.
   * @deprecated Use `handleInstancesPolymorphically`
   */
  arePolymorphic?: boolean;

  /**
   * Should instances be queried using a polymorphic query - from `classes` and all their
   * subclasses. This doesn't mean the resulting content will have all properties of the subclasses
   * though - they're only taken from base classes specified in `classes` attribute.
   */
  handleInstancesPolymorphically?: boolean;

  /**
   * Condition for filtering instances of defined classes.
   *
   * **See:** [ECExpressions available in instance filter]($docs/learning/presentation/Content/ECExpressions.md#instance-filter)
   */
  instanceFilter?: string;
}
