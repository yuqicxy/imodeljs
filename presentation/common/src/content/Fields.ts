/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Content
 */

import { ClassInfo, ClassInfoJSON, RelatedClassInfo, RelationshipPath, RelationshipPathJSON } from "../EC";
import { CategoryDescription } from "./Category";
import { EditorDescription } from "./Editor";
import { Property, PropertyJSON } from "./Property";
import { TypeDescription } from "./TypeDescription";

/**
 * Data structure for a [[Field]] serialized to JSON.
 * @internal
 */
export interface BaseFieldJSON {
  category: CategoryDescription;
  name: string;
  label: string;
  type: TypeDescription;
  isReadonly: boolean;
  priority: number;
  editor?: EditorDescription;
}

/**
 * Data structure for a [[PropertiesField]] serialized to JSON.
 * @internal
 */
export interface PropertiesFieldJSON extends BaseFieldJSON {
  properties: PropertyJSON[];
}

/**
 * Data structure for a [[NestedContentField]] serialized to JSON.
 * @internal
 */
export interface NestedContentFieldJSON extends BaseFieldJSON {
  contentClassInfo: ClassInfoJSON;
  pathToPrimaryClass: RelationshipPathJSON;
  autoExpand?: boolean;
  nestedFields: FieldJSON[];
}

/** @internal */
export type FieldJSON = BaseFieldJSON | PropertiesFieldJSON | NestedContentFieldJSON;

/**
 * Is supplied field a properties field.
 * @internal
 */
const isPropertiesField = (field: FieldJSON | Field): field is PropertiesFieldJSON | PropertiesField => {
  return (field as any).properties;
};

/**
 * Is supplied field a nested content field.
 * @internal
 */
const isNestedContentField = (field: FieldJSON | Field): field is NestedContentFieldJSON | NestedContentField => {
  return (field as any).nestedFields;
};

/**
 * Describes a single content field. A field is usually represented as a grid column
 * or a property pane row.
 *
 * @public
 */
export class Field {
  /** Category information */
  public category: CategoryDescription;
  /** Unique name */
  public name: string;
  /** Display label */
  public label: string;
  /** Description of this field's values data type */
  public type: TypeDescription;
  /** Are values in this field read-only */
  public isReadonly: boolean;
  /** Priority of the field. Higher priority fields should appear first in the UI */
  public priority: number;
  /** Property editor used to edit values of this field */
  public editor?: EditorDescription;
  /** Parent field */
  private _parent?: NestedContentField;

  /**
   * Creates an instance of Field.
   * @param category Category information
   * @param name Unique name
   * @param label Display label
   * @param type Description of this field's values data type
   * @param isReadonly Are values in this field read-only
   * @param priority Priority of the field
   * @param editor Property editor used to edit values of this field
   */
  public constructor(category: CategoryDescription, name: string, label: string, type: TypeDescription,
    isReadonly: boolean, priority: number, editor?: EditorDescription) {
    this.category = category;
    this.name = name;
    this.label = label;
    this.type = type;
    this.isReadonly = isReadonly;
    this.priority = priority;
    this.editor = editor;
  }

  /**
   * Is this a [[PropertiesField]]
   */
  public isPropertiesField(): this is PropertiesField { return isPropertiesField(this); }

  /**
   * Is this a [[NestedContentField]]
   */
  public isNestedContentField(): this is NestedContentField { return isNestedContentField(this); }

  /**
   * Get parent
   */
  public get parent(): NestedContentField | undefined { return this._parent; }

  /** @internal */
  public toJSON(): FieldJSON {
    return {
      category: this.category,
      name: this.name,
      label: this.label,
      type: this.type,
      isReadonly: this.isReadonly,
      priority: this.priority,
      editor: this.editor,
    };
  }

  /**
   * Deserialize Field from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized field or undefined if deserialization failed
   *
   * @internal
   */
  public static fromJSON(json: FieldJSON | string | undefined): Field | undefined {
    if (!json)
      return undefined;
    if (typeof json === "string")
      return JSON.parse(json, Field.reviver);
    if (isPropertiesField(json))
      return PropertiesField.fromJSON(json);
    if (isNestedContentField(json))
      return NestedContentField.fromJSON(json);
    const field = Object.create(Field.prototype);
    return Object.assign(field, json);
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing Field objects.
   *
   * @internal
   */
  public static reviver(key: string, value: any): any {
    return key === "" ? Field.fromJSON(value) : value;
  }

  /** @internal */
  public resetParentship(): void {
    this._parent = undefined;
  }

  /** @internal */
  public rebuildParentship(parentField?: NestedContentField): void {
    this._parent = parentField;
  }
}

/**
 * Describes a content field that's based on one or more similar
 * EC properties.
 *
 * @public
 */
export class PropertiesField extends Field {
  /** A list of properties this field is created from */
  public properties: Property[];

  /**
   * Creates an instance of PropertiesField.
   * @param category Category information
   * @param name Unique name
   * @param label Display label
   * @param type Description of this field's values data type
   * @param isReadonly Are values in this field read-only
   * @param priority Priority of the field
   * @param properties A list of properties this field is created from
   * @param editor Property editor used to edit values of this field
   */
  public constructor(category: CategoryDescription, name: string, label: string, description: TypeDescription,
    isReadonly: boolean, priority: number, properties: Property[], editor?: EditorDescription) {
    super(category, name, label, description, isReadonly, priority, editor);
    this.properties = properties;
  }

  /** @internal */
  public toJSON(): PropertiesFieldJSON {
    return {
      ...super.toJSON(),
      properties: this.properties.map((p) => Property.toJSON(p)),
    };
  }

  /**
   * Deserialize PropertiesField from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized properties field or undefined if deserialization failed
   *
   * @internal
   */
  public static fromJSON(json: PropertiesFieldJSON | string | undefined): PropertiesField | undefined {
    if (!json)
      return undefined;
    if (typeof json === "string")
      return JSON.parse(json, Field.reviver);
    const field = Object.create(PropertiesField.prototype);
    return Object.assign(field, json, {
      properties: json.properties.map((p) => Property.fromJSON(p)),
    } as Partial<PropertiesField>);
  }
}

/**
 * Describes a content field that contains [Nested content]($docs/learning/presentation/Content/Terminology#nested-content).
 *
 * @public
 */
export class NestedContentField extends Field {
  /** Information about an ECClass whose properties are nested inside this field */
  public contentClassInfo: ClassInfo;
  /** Relationship path to [Primary class]($docs/learning/presentation/Content/Terminology#primary-class) */
  public pathToPrimaryClass: RelationshipPath;
  /** Contained nested fields */
  public nestedFields: Field[];
  /** Flag specifying whether field should be expanded */
  public autoExpand?: boolean;

  /**
   * Creates an instance of NestedContentField.
   * @param category Category information
   * @param name Unique name
   * @param label Display label
   * @param type Description of this field's values data type
   * @param isReadonly Are values in this field read-only
   * @param priority Priority of the field
   * @param contentClassInfo Information about an ECClass whose properties are nested inside this field
   * @param pathToPrimaryClass Relationship path to [Primary class]($docs/learning/presentation/Content/Terminology#primary-class)
   * @param nestedFields Contained nested fields
   * @param autoExpand Flag specifying whether field should be expanded
   * @param editor Property editor used to edit values of this field
   */
  public constructor(category: CategoryDescription, name: string, label: string, description: TypeDescription,
    isReadonly: boolean, priority: number, contentClassInfo: ClassInfo, pathToPrimaryClass: RelationshipPath,
    nestedFields: Field[], editor?: EditorDescription, autoExpand?: boolean) {
    super(category, name, label, description, isReadonly, priority, editor);
    this.contentClassInfo = contentClassInfo;
    this.pathToPrimaryClass = pathToPrimaryClass;
    this.nestedFields = nestedFields;
    this.autoExpand = autoExpand;
  }

  /**
   * Get field by its name
   * @param name Name of the field to find
   * @param recurse Recurse into nested fields
   */
  public getFieldByName(name: string, recurse?: boolean): Field | undefined {
    return getFieldByName(this.nestedFields, name, recurse);
  }

  /** @internal */
  public toJSON(): NestedContentFieldJSON {
    return {
      ...super.toJSON(),
      contentClassInfo: this.contentClassInfo,
      pathToPrimaryClass: this.pathToPrimaryClass,
      nestedFields: this.nestedFields.map((field: Field) => field.toJSON()),
      autoExpand: this.autoExpand,
    };
  }

  /**
   * Deserialize NestedContentField from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized nested content field or undefined if deserialization failed
   *
   * @internal
   */
  public static fromJSON(json: NestedContentFieldJSON | string | undefined): NestedContentField | undefined {
    if (!json)
      return undefined;
    if (typeof json === "string")
      return JSON.parse(json, Field.reviver);
    const field = Object.create(NestedContentField.prototype);
    return Object.assign(field, json, {
      nestedFields: json.nestedFields.map((nestedFieldJson: FieldJSON) => Field.fromJSON(nestedFieldJson)).filter((nestedField) => (undefined !== nestedField)),
      contentClassInfo: ClassInfo.fromJSON(json.contentClassInfo),
      pathToPrimaryClass: json.pathToPrimaryClass.map((p) => RelatedClassInfo.fromJSON(p)),
      autoExpand: json.autoExpand,
    } as Partial<NestedContentField>);
  }

  /** @internal */
  public resetParentship(): void {
    super.resetParentship();
    for (const nestedField of this.nestedFields)
      nestedField.resetParentship();
  }

  /** @internal */
  public rebuildParentship(parentField?: NestedContentField): void {
    super.rebuildParentship(parentField);
    for (const nestedField of this.nestedFields)
      nestedField.rebuildParentship(this);
  }
}

/** @internal */
export const getFieldByName = (fields: Field[], name: string, recurse?: boolean): Field | undefined => {
  for (const field of fields) {
    if (field.name === name)
      return field;

    if (recurse && field.isNestedContentField()) {
      const nested = getFieldByName(field.nestedFields, name, recurse);
      if (nested)
        return nested;
    }
  }
  return undefined;
};
