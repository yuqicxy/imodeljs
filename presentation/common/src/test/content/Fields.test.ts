/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import * as faker from "faker";
import {
  createRandomECClassInfo, createRandomECClassInfoJSON,
  createRandomRelationshipPath, createRandomRelationshipPathJSON,
  createRandomCategory, createRandomPrimitiveTypeDescription,
  createRandomPrimitiveField, createRandomPrimitiveFieldJSON, createRandomNestedFieldJSON, createRandomNestedContentField,
} from "../_helpers/random";
import { PropertiesFieldJSON } from "../../content/Fields";
import { Field, PropertiesField, NestedContentField, PropertyValueFormat, StructTypeDescription, Property } from "../../presentation-common";

const generateTestData = () => {
  const testData: any = {};
  testData.baseFieldJSON = createRandomPrimitiveFieldJSON();
  testData.propertiesFieldJSON = {
    ...testData.baseFieldJSON,
    properties: [{
      property: {
        classInfo: createRandomECClassInfoJSON(),
        name: faker.random.word(),
        type: faker.database.type(),
      },
      relatedClassPath: createRandomRelationshipPathJSON(1),
    }],
  } as PropertiesFieldJSON;
  testData.nestedContentFieldJSON = createRandomNestedFieldJSON();
  return testData;
};

describe("Field", () => {

  describe("fromJSON", () => {

    let testData!: any;
    beforeEach(() => {
      testData = generateTestData();
    });

    it("creates valid Field from valid JSON", () => {
      const item = Field.fromJSON(testData.baseFieldJSON);
      expect(item).to.matchSnapshot();
    });

    it("creates valid PropertiesField from valid JSON", () => {
      const item = Field.fromJSON(testData.propertiesFieldJSON);
      expect(item).to.matchSnapshot();
    });

    it("creates valid NestedContentField from valid JSON", () => {
      const item = Field.fromJSON(testData.nestedContentFieldJSON);
      expect(item).to.matchSnapshot();
    });

    it("creates valid Field from valid serialized JSON", () => {
      const item = Field.fromJSON(JSON.stringify(testData.baseFieldJSON));
      expect(item).to.matchSnapshot();
    });

    it("returns undefined for undefined JSON", () => {
      const item = Field.fromJSON(undefined);
      expect(item).to.be.undefined;
    });

  });

  describe("isPropertiesField", () => {

    it("returns false for non-properties field", () => {
      const field = createRandomPrimitiveField();
      expect(!field.isPropertiesField());
    });

    it("returns true for properties field", () => {
      const property: Property = {
        property: {
          classInfo: createRandomECClassInfo(),
          name: faker.random.word(),
          type: faker.database.type(),
        },
        relatedClassPath: [],
      };
      const field = new PropertiesField(createRandomCategory(), faker.random.word(), faker.random.words(),
        createRandomPrimitiveTypeDescription(), faker.random.boolean(), faker.random.number(), [property]);
      expect(field.isPropertiesField());
    });

  });

  describe("isNestedContentField", () => {

    it("returns false for non-nested content field", () => {
      const field = createRandomPrimitiveField();
      expect(!field.isNestedContentField());
    });

    it("returns true for nested content field", () => {
      const field = new NestedContentField(createRandomCategory(), faker.random.word(), faker.random.words(),
        createRandomPrimitiveTypeDescription(), faker.random.boolean(), faker.random.number(), createRandomECClassInfo(),
        [], [], undefined, faker.random.boolean());
      expect(field.isNestedContentField());
    });

  });

});

describe("PropertiesField", () => {

  describe("fromJSON", () => {

    let testData!: any;
    beforeEach(() => {
      testData = generateTestData();
    });

    it("creates valid PropertiesField from valid JSON", () => {
      const item = PropertiesField.fromJSON(testData.propertiesFieldJSON);
      expect(item).to.matchSnapshot();
    });

    it("creates valid PropertiesField from valid serialized JSON", () => {
      const item = PropertiesField.fromJSON(JSON.stringify(testData.propertiesFieldJSON));
      expect(item).to.matchSnapshot();
    });

    it("returns undefined for undefined JSON", () => {
      const item = PropertiesField.fromJSON(undefined);
      expect(item).to.be.undefined;
    });

  });

});

describe("NestedContentField", () => {

  describe("getFieldByName", () => {

    it("returns undefined when there are no nested fields", () => {
      const field = createRandomNestedContentField([]);
      expect(field.getFieldByName("test")).to.be.undefined;
    });

    it("returns undefined when field is not found", () => {
      const field = createRandomNestedContentField();
      const name = field.nestedFields[0].name + "_does_not_exist";
      expect(field.getFieldByName(name, true)).to.be.undefined;
    });

    it("returns a field", () => {
      const field = createRandomNestedContentField();
      const nestedField = field.nestedFields[0];
      expect(field.getFieldByName(nestedField.name)).to.eq(nestedField);
    });

  });

  describe("fromJSON", () => {

    let testData!: any;
    beforeEach(() => {
      testData = generateTestData();
    });

    it("creates valid NestedContentField from valid JSON", () => {
      const item = NestedContentField.fromJSON(testData.nestedContentFieldJSON);
      expect(item).to.matchSnapshot();
    });

    it("creates valid NestedContentField from valid serialized JSON", () => {
      const item = NestedContentField.fromJSON(JSON.stringify(testData.nestedContentFieldJSON));
      expect(item).to.matchSnapshot();
    });

    it("creates valid NestedContentField from valid serialized JSON", () => {
      testData.nestedContentFieldJSON.nestedFields = [createRandomPrimitiveFieldJSON(), undefined as any];
      const field = NestedContentField.fromJSON(testData.nestedContentFieldJSON);
      expect(field!.nestedFields.length).to.eq(1);
    });

    it("returns undefined for undefined JSON", () => {
      const item = NestedContentField.fromJSON(undefined);
      expect(item).to.be.undefined;
    });

  });

  describe("rebuildParentship / resetParentship", () => {

    it("creates and resets parentship of self and nested fields", () => {
      const descr: StructTypeDescription = {
        valueFormat: PropertyValueFormat.Struct,
        typeName: faker.random.word(),
        members: [{
          type: createRandomPrimitiveTypeDescription(),
          label: faker.random.words(),
          name: faker.random.word(),
        }],
      };
      const field1 = createRandomPrimitiveField();
      const field2 = new NestedContentField(createRandomCategory(), faker.random.word(),
        faker.random.words(), descr, faker.random.boolean(), faker.random.number(),
        createRandomECClassInfo(), createRandomRelationshipPath(), [field1], undefined, faker.random.boolean());
      const field3 = new NestedContentField(createRandomCategory(), faker.random.word(),
        faker.random.words(), descr, faker.random.boolean(), faker.random.number(),
        createRandomECClassInfo(), createRandomRelationshipPath(), [field2], undefined, faker.random.boolean());

      field2.rebuildParentship(field3);
      expect(field3.parent).to.be.undefined;
      expect(field2.parent).to.eq(field3);
      expect(field1.parent).to.eq(field2);

      field3.resetParentship();
      expect(field3.parent).to.be.undefined;
      expect(field2.parent).to.be.undefined;
      expect(field1.parent).to.be.undefined;
    });

  });

});
