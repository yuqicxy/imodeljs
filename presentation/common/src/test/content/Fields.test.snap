// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field fromJSON creates valid Field from valid JSON 1`] = `
Object {
  "category": Object {
    "description": "Tenetur quo recusandae.",
    "expand": true,
    "label": "Oregon",
    "name": "Designer",
    "priority": 5601,
  },
  "editor": undefined,
  "isReadonly": true,
  "label": "Sleek SAS e-commerce",
  "name": "engineer",
  "priority": 59753,
  "type": Object {
    "typeName": "serial",
    "valueFormat": "Primitive",
  },
}
`;

exports[`Field fromJSON creates valid Field from valid serialized JSON 1`] = `
Object {
  "category": Object {
    "description": "Tempora eius eveniet vel.",
    "expand": false,
    "label": "Extension",
    "name": "auxiliary",
    "priority": 2776,
  },
  "editor": Object {
    "name": "JSON",
  },
  "isReadonly": true,
  "label": "Electronics tan",
  "name": "Garden",
  "priority": 81468,
  "type": Object {
    "typeName": "bit",
    "valueFormat": "Primitive",
  },
}
`;

exports[`Field fromJSON creates valid NestedContentField from valid JSON 1`] = `
Object {
  "autoExpand": false,
  "category": Object {
    "description": "Tempora consequuntur quam eveniet distinctio quos labore corporis neque distinctio.",
    "expand": false,
    "label": "multi-byte alarm content-based",
    "name": "withdrawal",
    "priority": 80265,
  },
  "contentClassInfo": Object {
    "id": "0x173a0000003faf",
    "label": "SSL",
    "name": "protocol",
  },
  "editor": undefined,
  "isReadonly": true,
  "label": "Handcrafted Sleek Sleek Frozen Sausages",
  "name": "solid state",
  "nestedFields": Array [
    Object {
      "category": Object {
        "description": "Dolores natus qui accusamus velit similique et dicta rerum nisi.",
        "expand": false,
        "label": "whiteboard Plains",
        "name": "Ferry",
        "priority": 49022,
      },
      "editor": undefined,
      "isReadonly": true,
      "label": "neural",
      "name": "Directives",
      "priority": 24586,
      "type": Object {
        "typeName": "date",
        "valueFormat": "Primitive",
      },
    },
  ],
  "pathToPrimaryClass": Array [
    Object {
      "isForwardRelationship": true,
      "isPolymorphicRelationship": false,
      "relationshipInfo": Object {
        "id": "0x17b830000012270",
        "label": "Ridges streamline",
        "name": "silver",
      },
      "sourceClassInfo": Object {
        "id": "0x1526a0000012015",
        "label": "matrix Focused Handmade Metal Hat",
        "name": "Kentucky",
      },
      "targetClassInfo": Object {
        "id": "0x10ff9000000ee9d",
        "label": "asynchronous Wooden",
        "name": "Ohio",
      },
    },
    Object {
      "isForwardRelationship": true,
      "isPolymorphicRelationship": false,
      "relationshipInfo": Object {
        "id": "0x13572000000f162",
        "label": "Soap",
        "name": "Iowa",
      },
      "sourceClassInfo": Object {
        "id": "0x4b30000000b1c3",
        "label": "algorithm Massachusetts",
        "name": "Niue",
      },
      "targetClassInfo": Object {
        "id": "0x149e900000066ed",
        "label": "Buckinghamshire architecture",
        "name": "South Carolina",
      },
    },
  ],
  "priority": 18043,
  "type": Object {
    "members": Array [
      Object {
        "label": "withdrawal",
        "name": "invoice",
        "type": Object {
          "typeName": "bigint",
          "valueFormat": "Primitive",
        },
      },
    ],
    "typeName": "Cambridgeshire",
    "valueFormat": "Struct",
  },
}
`;

exports[`Field fromJSON creates valid PropertiesField from valid JSON 1`] = `
Object {
  "category": Object {
    "description": "Corporis occaecati explicabo voluptatum ipsam debitis.",
    "expand": false,
    "label": "Sausages",
    "name": "structure",
    "priority": 59710,
  },
  "editor": Object {
    "name": "CSS",
  },
  "isReadonly": true,
  "label": "Toys feed platforms",
  "name": "Infrastructure",
  "priority": 72750,
  "properties": Array [
    Object {
      "property": Object {
        "classInfo": Object {
          "id": "0x12e040000011031",
          "label": "Savings Account Borders",
          "name": "Wooden",
        },
        "name": "Brand",
        "type": "binary",
      },
      "relatedClassPath": Array [
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": false,
          "relationshipInfo": Object {
            "id": "0xb220000013beb",
            "label": "Avon Plastic calculating",
            "name": "card",
          },
          "sourceClassInfo": Object {
            "id": "0x3b01000000711f",
            "label": "Chips hack",
            "name": "panel",
          },
          "targetClassInfo": Object {
            "id": "0x10040000006bba",
            "label": "plum Refined Steel Shoes",
            "name": "program",
          },
        },
      ],
    },
  ],
  "type": Object {
    "typeName": "int",
    "valueFormat": "Primitive",
  },
}
`;

exports[`NestedContentField fromJSON creates valid NestedContentField from valid JSON 1`] = `
Object {
  "autoExpand": true,
  "category": Object {
    "description": "Est fugit nisi occaecati ea eos maiores distinctio.",
    "expand": true,
    "label": "Borders",
    "name": "clicks-and-mortar",
    "priority": 65673,
  },
  "contentClassInfo": Object {
    "id": "0x1365a0000010344",
    "label": "Belize Dollar Bacon Cameroon",
    "name": "GB",
  },
  "editor": Object {
    "name": "Business-focused",
  },
  "isReadonly": true,
  "label": "compressing Metal",
  "name": "application",
  "nestedFields": Array [
    Object {
      "category": Object {
        "description": "Error delectus est ea omnis rerum perferendis optio voluptatibus.",
        "expand": false,
        "label": "Investor Designer",
        "name": "Colombian Peso Unidad de Valor Real",
        "priority": 98459,
      },
      "editor": undefined,
      "isReadonly": false,
      "label": "FTP",
      "name": "override",
      "priority": 56796,
      "type": Object {
        "typeName": "mediumint",
        "valueFormat": "Primitive",
      },
    },
  ],
  "pathToPrimaryClass": Array [
    Object {
      "isForwardRelationship": true,
      "isPolymorphicRelationship": false,
      "relationshipInfo": Object {
        "id": "0xd10e0000002a8d",
        "label": "granular",
        "name": "mesh",
      },
      "sourceClassInfo": Object {
        "id": "0x15498000000676f",
        "label": "Cambridgeshire Yemeni Rial functionalities",
        "name": "HTTP",
      },
      "targetClassInfo": Object {
        "id": "0xda2d000000bc8d",
        "label": "portals Kids",
        "name": "Tuna",
      },
    },
    Object {
      "isForwardRelationship": false,
      "isPolymorphicRelationship": false,
      "relationshipInfo": Object {
        "id": "0x43400000015918",
        "label": "EXE Credit Card Account",
        "name": "Montana",
      },
      "sourceClassInfo": Object {
        "id": "0x138c200000180f5",
        "label": "Frozen Arkansas calculating",
        "name": "Practical Metal Hat",
      },
      "targetClassInfo": Object {
        "id": "0x1275a00000169f4",
        "label": "indexing",
        "name": "Industrial",
      },
    },
  ],
  "priority": 50980,
  "type": Object {
    "members": Array [
      Object {
        "label": "calculate",
        "name": "programming",
        "type": Object {
          "typeName": "double",
          "valueFormat": "Primitive",
        },
      },
    ],
    "typeName": "North Carolina",
    "valueFormat": "Struct",
  },
}
`;

exports[`NestedContentField fromJSON creates valid NestedContentField from valid serialized JSON 1`] = `
Object {
  "autoExpand": false,
  "category": Object {
    "description": "Et rerum saepe odio quidem incidunt blanditiis.",
    "expand": true,
    "label": "primary 1080p",
    "name": "Diverse",
    "priority": 34887,
  },
  "contentClassInfo": Object {
    "id": "0x1010c0000004a68",
    "label": "Customer",
    "name": "Up-sized",
  },
  "editor": Object {
    "name": "Administrator",
  },
  "isReadonly": true,
  "label": "Tala",
  "name": "salmon",
  "nestedFields": Array [
    Object {
      "category": Object {
        "description": "Omnis exercitationem ut rerum sunt quo iste.",
        "expand": true,
        "label": "Mews content",
        "name": "seize",
        "priority": 4186,
      },
      "editor": undefined,
      "isReadonly": true,
      "label": "Chief Street",
      "name": "Shoes",
      "priority": 62421,
      "type": Object {
        "typeName": "bit",
        "valueFormat": "Primitive",
      },
    },
  ],
  "pathToPrimaryClass": Array [
    Object {
      "isForwardRelationship": false,
      "isPolymorphicRelationship": false,
      "relationshipInfo": Object {
        "id": "0xf4ff0000013d2a",
        "label": "protocol",
        "name": "Practical",
      },
      "sourceClassInfo": Object {
        "id": "0x142c90000011ff8",
        "label": "Kentucky protocol Handcrafted",
        "name": "CSS",
      },
      "targetClassInfo": Object {
        "id": "0x76000001437f",
        "label": "AI Fantastic Metal Computer",
        "name": "instruction set",
      },
    },
    Object {
      "isForwardRelationship": false,
      "isPolymorphicRelationship": false,
      "relationshipInfo": Object {
        "id": "0xa5e10000008592",
        "label": "Sudanese Pound",
        "name": "users",
      },
      "sourceClassInfo": Object {
        "id": "0x15fe00000000bdf",
        "label": "white",
        "name": "Wallis and Futuna",
      },
      "targetClassInfo": Object {
        "id": "0xe3bf0000011f42",
        "label": "Stravenue white",
        "name": "customized",
      },
    },
  ],
  "priority": 5085,
  "type": Object {
    "members": Array [
      Object {
        "label": "SAS",
        "name": "Cotton",
        "type": Object {
          "typeName": "binary",
          "valueFormat": "Primitive",
        },
      },
    ],
    "typeName": "Money Market Account",
    "valueFormat": "Struct",
  },
}
`;

exports[`PropertiesField fromJSON creates valid PropertiesField from valid JSON 1`] = `
Object {
  "category": Object {
    "description": "Sequi repudiandae velit atque voluptas.",
    "expand": true,
    "label": "Grocery West Virginia",
    "name": "Incredible",
    "priority": 39804,
  },
  "editor": undefined,
  "isReadonly": false,
  "label": "Tanzanian Shilling front-end optical",
  "name": "Producer",
  "priority": 57252,
  "properties": Array [
    Object {
      "property": Object {
        "classInfo": Object {
          "id": "0x49c00000089c8",
          "label": "Ferry enable Wooden",
          "name": "Arkansas",
        },
        "name": "SQL",
        "type": "mediumint",
      },
      "relatedClassPath": Array [
        Object {
          "isForwardRelationship": false,
          "isPolymorphicRelationship": true,
          "relationshipInfo": Object {
            "id": "0x1da9000000c48c",
            "label": "calculating Creek Forks",
            "name": "Syrian Arab Republic",
          },
          "sourceClassInfo": Object {
            "id": "0x17efb000000da5f",
            "label": "online",
            "name": "Indiana",
          },
          "targetClassInfo": Object {
            "id": "0x1220e000000b42b",
            "label": "Iraqi Dinar Mexican Peso Mexican Unidad de Inversion (UDI) Tasty",
            "name": "Small",
          },
        },
      ],
    },
  ],
  "type": Object {
    "typeName": "boolean",
    "valueFormat": "Primitive",
  },
}
`;

exports[`PropertiesField fromJSON creates valid PropertiesField from valid serialized JSON 1`] = `
Object {
  "category": Object {
    "description": "Vitae deleniti eum aperiam quo recusandae doloremque voluptatum alias ut.",
    "expand": false,
    "label": "Incredible Granite Towels wireless info-mediaries",
    "name": "orange",
    "priority": 59789,
  },
  "editor": Object {
    "name": "Dynamic",
  },
  "isReadonly": false,
  "label": "Investment Account copy bypassing",
  "name": "Garden",
  "priority": 32204,
  "properties": Array [
    Object {
      "property": Object {
        "classInfo": Object {
          "id": "0x801e00000056e7",
          "label": "Venezuela deposit Seamless",
          "name": "Fiji Dollar",
        },
        "name": "Cambridgeshire",
        "type": "double",
      },
      "relatedClassPath": Array [
        Object {
          "isForwardRelationship": true,
          "isPolymorphicRelationship": false,
          "relationshipInfo": Object {
            "id": "0xb447000000b6b8",
            "label": "Rustic Plastic Shoes e-commerce",
            "name": "front-end",
          },
          "sourceClassInfo": Object {
            "id": "0xce2800000043fa",
            "label": "Central",
            "name": "haptic",
          },
          "targetClassInfo": Object {
            "id": "0xf9330000009885",
            "label": "Handmade",
            "name": "Health",
          },
        },
      ],
    },
  ],
  "type": Object {
    "typeName": "binary",
    "valueFormat": "Primitive",
  },
}
`;
